[WIP] Add parquet support

zhkvia opened the pull request:
In this PR [Parquet](https://parquet.apache.org/) data format support is implemented.
The input and output processing is implemented using [parquet-cpp](https://github.com/apache/parquet-cpp/) library in conjunction with [arrow](https://github.com/apache/arrow).

Current state:
- New `Format` is created — `Parquet`.  Common usages would be:
    Sending data from a file to a server:
    ```bash
    $ cat data.parquet | clickhouse-client --query="INSERT INTO table FORMAT Parquet"
    ```

    Storing table data into a parquet file:
    ```bash
    $ clickhouse-client --query="SELECT * FROM table FORMAT Parquet" > data.parquet
    ```
- Supported input [types](https://github.com/apache/parquet-format/blob/master/LogicalTypes.md)* (Parquet type -> ClickHouse type):
    - [x] All numerical types (`INT_8`, `INT_16`, `INT_32`, `INT_64`, `UINT_8`, `UINT_16`, `UINT_32`, `UINT_64`, `FLOAT`, `DOUBLE`)
    - [x] `BOOLEAN` -> `UInt8`
    - [x] `DATE` -> `Date` (**Note:** **arrow** library stores `DATE` type as Int32, while ClickHouse stores `Date` as UInt16. Therefore implicit conversion is performed. Also internal dates are limited by the compilation-time variable [DATE_LUT_MAX_DAY_NUM](https://github.com/yandex/ClickHouse/blob/master/libs/libcommon/include/common/DateLUTImpl.h#L10))
    - [x] `UTF8` -> `String`
    TODO?:
    - [ ] `UUID` -> `String`
    - [ ] `JSON` -> `String`
    - [ ] `TIME_MILLIS`
    - [ ] `TIME_MICROS`
    - [ ] Nested types
- Supported output [types](https://github.com/apache/parquet-format/blob/master/LogicalTypes.md)* (ClickHouse type -> Parquet type):
    - [x] All numerical types (`INT_8`, `INT_16`, `INT_32`, `INT_64`, `UINT_8`, `UINT_16`, `UINT_32`, `UINT_64`, `FLOAT`, `DOUBLE`)
    - [x] `Date` -> `DATE`
    - [x] `String` -> `UTF8`
    - [ ] `FixedString` -> `BYTE_ARRAY`
    - [ ] `DateTime` -> `TIMESTAMP_MILLIS`
  TODO?:
    - [ ] Nested types

&ast; Types that are not included are probably not convertable into internal (ClickHouse) ones.

TODO:
- [x] Support NULLs
~~- [ ] Optimize file reading (do not store a whole file in memory)~~ **UPD:** Seems to be impossible since metadata is stored at the end of a data stream
- [ ] Output a file using [row group](https://parquet.apache.org/documentation/latest/) size estimation (a number of rows in a continuous chunk of data)
- [x] Make parquet support optional on a build step
- [x] Set up library building within the project using CMake
- [x] Use compression
- [ ] Write tests
- [ ] Add docs

---
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

zhkvia committed:
Add a parquet-cpp repo to contrib

zhkvia committed:
Add cmake modules for libraries detection

zhkvia committed:
Link parquet and arrow libraries

zhkvia committed:
Add a template for ParquetBlockInputStream

zhkvia committed:
Parquet input: support numeric types

zhkvia committed:
Parquet input: support a string type

zhkvia committed:
Parquet input: get rid of redundant copying

zhkvia committed:
Parquet input: add a boolean type support

zhkvia committed:
Parquet input: change if to switch

zhkvia committed:
Parquet input: add a Date type support

zhkvia committed:
Parquet output: support numeric and string types

zhkvia committed:
Check arrow status

zhkvia committed:
Parquet output: refactor functions-fillers

zhkvia committed:
Actualize TODOs

ludv1x commented on dbms/src/DataStreams/ParquetBlockInputStream.h:
```diff
@@ -0,0 +1,62 @@
+#include <DataTypes/DataTypeString.h>
+#include <DataTypes/DataTypeDate.h>
+// TODO: refine includes
+#include <arrow/api.h>
```
It is better not to show symbols from externals libraries in the header file.

ludv1x commented on dbms/src/DataStreams/ParquetBlockOutputStream.cpp:
```diff
@@ -0,0 +1,181 @@
+        arrow_fields.emplace_back(new arrow::Field(column.name, internal_type_to_arrow_type[column.type->getName()], /*nullable = */false));
+        std::shared_ptr<arrow::Array> arrow_array;
+
+        String internal_type_name = column.type->getName();
```
Use typeid_cast instead of String comparison to dispatch types.

ludv1x commented on dbms/src/DataStreams/ParquetBlockOutputStream.h:
```diff
@@ -0,0 +1,52 @@
+    void fillArrowArrayWithDateColumnData(ColumnPtr write_column, std::shared_ptr<arrow::Array> & arrow_array);
+    void fillArrowArrayWithStringColumnData(ColumnPtr write_column, std::shared_ptr<arrow::Array> & arrow_array);
+
+    std::unordered_map<String, std::shared_ptr<arrow::DataType>> internal_type_to_arrow_type = {
```
It looks like it should be `static const`.
And defined in cpp.

ludv1x commented on dbms/src/DataStreams/ParquetBlockOutputStream.h:
```diff
@@ -0,0 +1,52 @@
+
+        {"Date",  arrow::date32()},
+
+        {"String", arrow::utf8()}//,
```
I am not sure about it. ClickHouse String could store strings that are not valid utf-8 sequences.

zhkvia committed:
Parquet input: support NULL values

zhkvia committed:
Refactor: make some members and methods static

maddhuapplicate commented:
Any idea when this would be released?

alexey-milovidov commented:
@luc1ph3r will try to prepare this PR for merge in September.

zhkvia committed:
Update submodules since parquet is a part of arrow

- https://github.com/apache/parquet-cpp/commit/f84dca4203930fa2898d0d98d278d74520fafec4

zhkvia committed:
Parquet output: support NULL values

zhkvia committed:
Parquet data format: refactoring

zhkvia committed:
Parquet data format: update CMake configurations

arthurdk commented:
Hello, are you still planning on working on that feature? It would be really awesome :smiley:

blinkov commented:
@luc1ph3r any news?

alexey-milovidov commented:
Superseded by #3759.
